usr/src/test/zfs-tests/tests/functional/snapshot/deadlist_lock.ksh

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/ksh -p
+#
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright (c) 2016 by Delphix. All rights reserved.
+#
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/snapshot/snapshot.cfg
+
+#
+# DESCRIPTION:
+#
+# This test ensures that the following race condition does not
+# take place:
+#
+# 1] A sync thread inserts a new entry in the deadlist of a
+#    snapshot. The dle_bpobj at that entry currently is the
+#    empty bpobj (our sentinel), so we close it and we are
+#    about to reopen it. (see dle_enqueue())
+#
+# 2] At the same time a thread executing an administrative
+#    command that uses dsl_deadlist_space_range() is about
+#    to dereference that same bpobj that was just closed
+#    and therefore is NULL.
+#
+# 3] The sync thread loses the race and we dereference the
+#    NULL pointer in the kernel.
+#
+# STRATEGY:
+#
+# 1. Setup a folder and create a bunch of test files. Take a
+#    snapshot right after you create a new test file.
+# 2. Start DTrace in the background to put a delay in the
+#    sync thread after it closes the empty bpobj and before
+#    it reopens it.
+# 3. Start a process in the backgroud that runs zfs-destroy
+#    dry-runs in an infinite loop. The idea is to keep calling
+#    dsl_deadlist_space_range().
+# 4. Go ahead and start removing the test files. This should
+#    start populating the deadlist of each snapshot with
+#    entries and go through the dle_enqueue() target code.
+# 5. If the test passes, kill the process running on a loop
+#    and dtrace, and cleanup the dataset.
+#
+
+verify_runnable "both"
+
+
+DLDS="dl_race"
+
+function cleanup
+{
+	log_must kill -9 $DLOOP_PID $DTRACE_PID
+	log_must zfs destroy -fR $TESTPOOL/$TESTFS/$DLDS
```

Comment:
We now have a different method for finding the utilities, we modify the `$PATH` to contain only a directory that has symlinks to the utilities.  This was changed by the following illumos commit:
```
commit 1d32ba663e202c24a5a1f2e5aef83fffb447cb7f
Author: John Wren Kennedy <john.kennedy@delphix.com>
Date:   Fri Nov 11 09:57:19 2016 -0800

    7290 ZFS test suite needs to control what utilities it can run
```